This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [6]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case
API_KEY = '#######'

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [1]:
# First, import the relevant modules
import requests


In [30]:
url_check = 'https://www.quandl.com/api/v3/datasets/FSE/AFX_X/data.json?start_date=2017-01-01&end_date=2017-01-02api_key=#######'

In [31]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
check = requests.get(url_check)

In [32]:
print(check)
print(check.text)

<Response [200]>
{"dataset_data":{"limit":null,"transform":null,"column_index":null,"column_names":["Date","Open","High","Low","Close","Change","Traded Volume","Turnover","Last Price of the Day","Daily Traded Units","Daily Turnover"],"start_date":"2017-01-01","end_date":"2017-01-02","frequency":"daily","data":[["2017-01-02",34.99,35.94,34.99,35.8,null,44700.0,1590561.0,null,null,null]],"collapse":null,"order":null}}


In [34]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
json_data_check = check.json()
    
for k in json_data_check.keys():
    print(k + ': ', json_data_check[k])

dataset_data:  {'limit': None, 'transform': None, 'column_index': None, 'column_names': ['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover'], 'start_date': '2017-01-01', 'end_date': '2017-01-02', 'frequency': 'daily', 'data': [['2017-01-02', 34.99, 35.94, 34.99, 35.8, None, 44700.0, 1590561.0, None, None, None]], 'collapse': None, 'order': None}


These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

Task 1 Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).

In [38]:
url = 'https://www.quandl.com/api/v3/datasets/FSE/AFX_X/data.json?start_date=2017-01-01&end_date=2017-12-31api_key=SzBzFWKXfdXGzW3xz6dz'

In [39]:
AFX_X = requests.get(url)

In [40]:
print(AFX_X)

<Response [200]>


Task 2 Convert the returned JSON object into a Python dictionary.

In [111]:
AFX_X_json = AFX_X.json()
    
'''for k in AFX_X_json.keys():
    print(k + ': ', AFX_X_json[k]);'''

"for k in AFX_X_json.keys():\n    print(k + ': ', AFX_X_json[k]);"

In [43]:
type(AFX_X_json)

dict

Task 3 Calculate what the highest and lowest opening prices were for the stock in this period.

In [121]:
values = AFX_X_json['dataset_data']['data']
AFX_X_prices = {item[0]: item[1:] for item in values}
AFX_X_open = {item[0]: item[1] for item in values}
AFX_X_high = {item[0]: item[2] for item in values}
AFX_X_low = {item[0]: item[3] for item in values}
AFX_X_close = {item[0]: item[4] for item in values}
AFX_X_volume = {item[0]: item[6] for item in values}

'''print(AFX_X_open)'''


'print(AFX_X_open)'

Create dictionary = {date:[stock info]
Important keys: Date=0, Open=1, High=2, Low=3, Close=4, Volume=6


Code modified from: [https://stackoverflow.com/questions/34191817/python-list-of-lists-to-dictionary]

In [128]:
'''print(AFX_X_open)'''
filtered = {k: v for k, v in AFX_X_open.items() if v is not None}
AFX_X_open.clear()
AFX_X_open.update(filtered)
print(max(AFX_X_open.values()))
print(min(AFX_X_open.values()))

53.11
34.0


Code modified from: https://stackoverflow.com/questions/33797126/proper-way-to-remove-keys-in-dictionary-with-none-values-in-python

Answer: The highest opening price is $53.11 and the lowest opening price is $34.00

Task 4 What was the largest change in any one day (based on High and Low price)?

In [132]:
day_range = {x: AFX_X_high[x] - AFX_X_low[x] for x in AFX_X_high if x in AFX_X_low}
max_range = max(day_range.values())
max_keys = [k for k, v in day_range.items() if v == max_range]
print(max_keys, max_range)

['2017-05-11'] 2.8100000000000023


Code modified from: https://stackoverflow.com/questions/43371951/getting-the-difference-in-values-between-two-dictionaries-in-python, https://stackoverflow.com/questions/42044090/return-the-maximum-value-from-a-dictionary/42044202

Answer: The largest change in any one day was $2.81 on May 11, 2017

Task 5 What was the largest change between any two days (based on Closing Price)?

In [136]:
filtered = {k: v for k, v in AFX_X_close.items() if v is not None}
AFX_X_close.clear()
AFX_X_close.update(filtered)
max_close = max(AFX_X_close.values())
max_close_key = [k for k, v in AFX_X_close.items() if v == max_close]
min_close = min(AFX_X_close.values())
min_close_key = [k for k, v in AFX_X_close.items() if v == min_close]
close_range = max_close - min_close
print(min_close_key, min_close, max_close_key, max_close, close_range)

['2017-01-23'] 34.06 ['2017-12-13'] 53.09 19.03


Answer: The largest close change between any two days is $19.03. This range is from the high close of $53.09 on December 13th, 2017 and the low close of $34.06 on January 23rd, 2017

Task 6 What was the average daily trading volume during this year?

In [139]:
average_volume = sum(AFX_X_volume.values()) / len(AFX_X_volume)
  
print(average_volume)

89124.33725490196


Answer: The average daily volume is 89,124.34

Task 7 (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [148]:
volume_list = list(AFX_X_volume.values())
volume_list.sort()
median = volume_list[len(volume_list) // 2]
print(median)


76286.0


Idea for method from: https://ucsb-cs8-w18-matni.github.io/lectures/CS8_Lecture11.pdf

Answer: The median volume is 76286.0